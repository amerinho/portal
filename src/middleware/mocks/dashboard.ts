export const dashboardSample = {
	value: [
		{
			address: { postalCode: "D00 001", streetAddress: "10 Westland Square, Pearse Street, Dublin 2" },
			id: "1",
			identifier: [{ valueReference: "Active", value: "ABC123" }]
		}
	]
};
