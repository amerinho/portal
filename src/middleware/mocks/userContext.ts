export const userContext = {
	b2c: {
		id: "4af3dfb6-2df6-448e-a5e9-4edc4323e033",
		displayName: "QA test",
		email: "QAtest01@parse.meetinireland.com"
	},
	crm: {
		contact: {
			id: "6148a839-05ef-ed11-9135-005056a0ce81",
			fullName: "John Doe"
		},
		accounts: [
			{
				id: "6248a839-05ef-ed11-9135-005056a0ce81",
				name: "Testing again",
				number: "FI037298",
				roles: [
					{
						name: "Financial Contact"
					},
					{
						name: "Trade Contact"
					},
					{
						name: "Premise Contact"
					}
				]
			}
		]
	}
};
