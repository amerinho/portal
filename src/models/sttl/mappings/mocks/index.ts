export * from "./formState";
export * from "./order";
export * from "./userAccount";
