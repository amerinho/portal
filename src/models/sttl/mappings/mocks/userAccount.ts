export const userAccount = {
	userId: "e86b4ebd-1a38-4757-a225-30fc2a2fc0e3",
	id: "e86b4ebd-1a38-4757-a225-30fc2a2fc0e3",
	provider: "https://fidigitalenterprisedps.b2clogin.com/6548f36a-b4b0-4666-9615-e4172baf2f4d/v2.0/",
	isNewUser: false,
	name: "QA test",
	email: "QAtest01@parse.meetinireland.com",
	firstName: "QA",
	lastName: "test"
};
