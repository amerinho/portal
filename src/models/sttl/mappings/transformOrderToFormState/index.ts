export * from "./transformOrderToFormState";
export * from "./isValidOrder";
export * from "./utils";
