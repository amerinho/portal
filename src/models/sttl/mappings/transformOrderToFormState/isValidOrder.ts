import { PostalAddress, ContainsPlaceItemForRooms, EntirePropertyOwns, MultipleUnitsOwns, OfferItem, Order, RoomsOwns } from "../../types/order";

type PermissionProperty = {
	"@type": string;
	name: string;
	value: PermissionStatus;
};
type QuantitativeValue = {
	"@type": "QuantitativeValue";
	unitText: string;
	value: string;
};

type LocationFeatureSpecification = {
	"@type": "LocationFeatureSpecification";
	name: string;
	value: string;
};

export const isValidOrder = (obj: any): obj is Order => {
	const order =
		typeof obj === "object" &&
		typeof obj["@context"] === "string" &&
		obj["@type"] === "Order" &&
		Array.isArray(obj.acceptedOffer) &&
		obj.acceptedOffer?.every(offer => isOfferItem(offer)) &&
		isCustomer(obj.customer) &&
		isPartOfInvoice(obj.partOfInvoice);
	return order;
};

export const isCustomer = (obj: any): obj is Order["customer"] => {
	const customer =
		typeof obj === "object" &&
		typeof obj["@type"] === "string" &&
		typeof obj.email === "string" &&
		typeof obj.name === "string" &&
		typeof obj.telephone === "string" &&
		Array.isArray(obj.identifier) &&
		obj.identifier.every(
			(id: any) =>
				typeof id === "object" &&
				typeof id["@type"] === "string" &&
				typeof id.propertyID === "string" &&
				typeof id.value === "string" &&
				typeof id.valueReference === "string"
		) &&
		typeof obj.worksFor === "object" &&
		typeof obj.worksFor["@type"] === "string" &&
		typeof obj.worksFor.name === "string";
	return customer;
};

export const isPartOfInvoice = (obj: any): obj is Order["partOfInvoice"] => {
	const partOfInvoice =
		typeof obj === "object" &&
		typeof obj["@type"] === "string" &&
		typeof obj.totalPaymentDue === "object" &&
		typeof obj.totalPaymentDue["@type"] === "string" &&
		typeof obj.totalPaymentDue.priceCurrency === "string" &&
		(typeof obj.totalPaymentDue.price === "number" || typeof obj.totalPaymentDue.price === "undefined");
	return partOfInvoice;
};

export const isOfferItem = (obj: any): obj is OfferItem => {
	const offerItem =
		typeof obj === "object" &&
		obj?.["@type"] === "Offer" &&
		obj?.priceCurrency === "EUR" &&
		obj?.itemOffered?.["@type"] === "GovernmentService" &&
		obj?.itemOffered?.name === "STL Registration" &&
		obj?.itemOffered?.isRelatedTo?.["@type"] === "Person" &&
		typeof obj?.itemOffered?.isRelatedTo?.name === "string" &&
		typeof obj?.itemOffered?.isRelatedTo?.givenName === "string" &&
		typeof obj?.itemOffered?.isRelatedTo?.familyName === "string" &&
		typeof obj?.itemOffered?.isRelatedTo?.email === "string" &&
		typeof obj?.itemOffered?.isRelatedTo?.telephone === "string" &&
		typeof obj?.itemOffered?.isRelatedTo?.address === "object" &&
		typeof obj.itemOffered.isRelatedTo.address.streetAddress === "string" &&
		typeof obj.itemOffered.isRelatedTo.address.postalCode === "string" &&
		typeof obj.itemOffered.isRelatedTo.address.addressLocality === "string" &&
		typeof obj.itemOffered.isRelatedTo.address.addressRegion === "string" &&
		typeof obj.itemOffered.isRelatedTo.address.addressCountry === "string" &&
		obj?.itemOffered?.isRelatedTo?.worksFor?.["@type"] === "Organization" &&
		typeof obj?.itemOffered?.isRelatedTo?.worksFor?.name === "string" &&
		typeof obj.itemOffered.isRelatedTo.owns === "object" &&
		obj.itemOffered.isRelatedTo.owns.additionalProperty?.every((item: PermissionProperty) => isPermissionStatus(item.value)) &&
		(isRoomsOwns(obj.itemOffered.isRelatedTo.owns) ||
			isEntirePropertyOwns(obj.itemOffered.isRelatedTo.owns) ||
			isMultipleUnitsOwns(obj.itemOffered.isRelatedTo.owns));

	return offerItem;
};

const isPermissionStatus = (value: any): value is PermissionStatus => value === "Granted" || value === "Outstanding";

const isLocationFeatureSpecification = (feature: any): feature is LocationFeatureSpecification =>
	typeof feature === "object" &&
	feature["@type"] === "LocationFeatureSpecification" &&
	typeof feature.name === "string" &&
	typeof feature.value === "string";

const isQuantitativeValue = (value: any): value is QuantitativeValue =>
	typeof value === "object" && value !== null && value["@type"] === "QuantitativeValue" && typeof value.unitText === "string";

const isContainsPlaceItemForRooms = (item: any): item is ContainsPlaceItemForRooms =>
	typeof item === "object" &&
	item !== null &&
	typeof item.additionalType === "string" &&
	typeof item.numberOfRooms === "object" &&
	isQuantitativeValue(item.numberOfRooms) &&
	Array.isArray(item.amenityFeature) &&
	item.amenityFeature.every(isLocationFeatureSpecification);

const isRoomsOwns = (obj: any): obj is RoomsOwns => {
	return (
		typeof obj === "object" &&
		typeof obj["@type"] === "string" &&
		isAddress(obj.address) &&
		Array.isArray(obj.containsPlace) &&
		obj.containsPlace.every((item: any) => isContainsPlaceItemForRooms(item))
	);
};

const isEntirePropertyOwns = (obj: any): obj is EntirePropertyOwns => {
	return (
		typeof obj === "object" &&
		typeof obj["@type"] === "string" &&
		isAddress(obj.address) &&
		typeof obj.numberOfBedrooms === "string" &&
		Array.isArray(obj.amenityFeature) &&
		obj.amenityFeature.length > 0 &&
		obj.amenityFeature[0]["@type"] === "LocationFeatureSpecification" &&
		obj.amenityFeature[0].name === "occupancy" &&
		typeof obj.amenityFeature[0].value === "string"
	);
};

const isMultipleUnitsOwns = (obj: any): obj is MultipleUnitsOwns => {
	return obj != null && typeof obj === "object" && typeof obj["@type"] === "string" && isAddress(obj.address);
};

export const isAddress = (obj: any): obj is PostalAddress => {
	return (
		typeof obj === "object" &&
		obj["@type"] === "PostalAddress" &&
		typeof obj.streetAddress === "string" &&
		typeof obj.postalCode === "string" &&
		typeof obj.addressLocality === "string" &&
		typeof obj.addressRegion === "string" &&
		typeof obj.addressCountry === "string"
	);
};
