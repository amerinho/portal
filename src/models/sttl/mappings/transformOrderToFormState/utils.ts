import { PROPERTY_OPTIONS, TEXT_TO_UNIT_ROOM_TYPE } from "settings/propertyTypeOptions";
import { RoomsOwns, EntirePropertyOwns, MultipleUnitsOwns, OfferItem, AdditionalProperty } from "../../types/order";
import { TEXT_TO_CATEGORY, Category, TEXT_TO_PLANNING_PERMISSION } from "../../types";

export const getPlanningPermissionFromOfferItem = (item: OfferItem) => {
	const permission = item.itemOffered.isRelatedTo.owns.additionalProperty?.find(ap => ap.name === "Planning Permission");
	return { permissionStatus: TEXT_TO_PLANNING_PERMISSION[permission.value] ?? "" };
};
export const getPlanningPermission = (additionalProperty: AdditionalProperty) => {
	const value = additionalProperty?.find(ap => ap.name === "Planning Permission")?.value;
	return TEXT_TO_PLANNING_PERMISSION[value] ?? "";
};

export const getPropertyType = (item: OfferItem) => {
	const {
		itemOffered: { category: categoryAsText }
	} = item;

	const category = TEXT_TO_CATEGORY[categoryAsText];
	const stepOneArrayItem = { category };

	if (category === Category.room) {
		const owns = item.itemOffered.isRelatedTo.owns as RoomsOwns;
		const { containsPlace } = owns;
		const hasRoomProperty = PROPERTY_OPTIONS.room.some(({ value }) => value === owns.additionalType);
		const propertyType = hasRoomProperty ? owns.additionalType : "Other - specify";
		const customPropertyType = hasRoomProperty ? undefined : owns.additionalType;

		const baseProps = { propertyType, ...(customPropertyType ? { customPropertyType } : {}) };

		const propertyDetails = containsPlace.reduce(
			(acc, next) => {
				const numberOfRooms = +next.numberOfRooms.value;
				const numberOfGuests = +next.amenityFeature[0].value;
				if (next.additionalType === "shared") {
					acc["numberOfSharedRooms"] = numberOfRooms;
					acc["numberOfGuestsInSharedRooms"] = numberOfGuests;
					return acc;
				}
				if (next.additionalType === "private") {
					acc["numberOfPrivateRooms"] = numberOfRooms;
					acc["numberOfGuestsInPrivateRooms"] = numberOfGuests;
					return acc;
				}
				return acc;
			},
			{ ...baseProps }
		);

		stepOneArrayItem[category] = propertyDetails;
		return stepOneArrayItem;
	}

	if (category === Category.fullProperty) {
		const owns = item.itemOffered.isRelatedTo.owns as EntirePropertyOwns;
		const hasFullProperty = PROPERTY_OPTIONS.fullProperty.some(({ value }) => value === owns.additionalType);
		const propertyType = hasFullProperty ? owns.additionalType : "Other - specify";
		const customPropertyType = hasFullProperty ? undefined : owns.additionalType;

		const baseProps = { propertyType, ...(customPropertyType ? { customPropertyType } : {}) };

		const propertyDetails = {
			...baseProps,
			numberOfBedrooms: +owns.numberOfBedrooms,
			numberOfGuests: +owns.amenityFeature[0].value
		};
		stepOneArrayItem[category] = propertyDetails;
		return stepOneArrayItem;
	}

	if (category === Category.units) {
		const owns = item.itemOffered.isRelatedTo.owns as MultipleUnitsOwns;

		const propertyDetails = owns.containsPlace.reduce((acc, next) => {
			const unitText = next.numberOfRooms?.unitText;
			const numberOfRooms = next.numberOfRooms?.value ? +next.numberOfRooms.value : 0;
			const numberOfGuests = next.amenityFeature[0]?.value ? +next.amenityFeature[0].value : 0;
			const hasUnitsProperty = PROPERTY_OPTIONS.units.some(({ value }) => value === next.additionalType);
			const propertyType = hasUnitsProperty ? next.additionalType : "Other - specify";
			const customPropertyType = hasUnitsProperty ? undefined : next.additionalType;

			const baseProps = { propertyType, ...(customPropertyType ? { customPropertyType } : {}) };

			const item = { ...baseProps, noOfGuests: numberOfGuests };
			if (unitText) {
				const prop = TEXT_TO_UNIT_ROOM_TYPE[unitText];
				item[prop] = numberOfRooms;
			}
			acc.push(item);
			return acc;
		}, []);

		stepOneArrayItem[category] = propertyDetails;
		return stepOneArrayItem;
	}
};

export const getPropertyAddress = (item: OfferItem) => {
	const { addressLocality, addressRegion, postalCode, streetAddress } = item.itemOffered.isRelatedTo.owns.address;

	const propertyAddress = {
		addressLine1: streetAddress,
		addressLine2: "", // will never know
		addressLine3: "", // will never know
		town: addressLocality,
		postcode: postalCode,
		county: `Co. ${addressRegion}`
	};
	return { propertyAddress };
};

export const getOwnerDetails = (item: OfferItem) => {
	const { address } = item.itemOffered.isRelatedTo;
	const businessName = item.itemOffered.isRelatedTo?.worksFor?.name || "";
	const { owns } = item.itemOffered.isRelatedTo;
	const ownsAddress = owns.address;
	const { givenName, familyName, email, telephone } = item.itemOffered.isRelatedTo;
	const isAddressSameAsStlProperty = address.streetAddress.trim() === ownsAddress.streetAddress.trim();
	if (isAddressSameAsStlProperty)
		return {
			countryOfResidence: address.addressCountry,
			firstName: givenName,
			lastName: familyName,
			emailAddress: email,
			telephone,
			businessName,
			isAddressSameAsStlProperty: true
		};
	return {
		countryOfResidence: address.addressCountry,
		firstName: givenName,
		lastName: familyName,
		emailAddress: email,
		telephone,
		businessName,
		isAddressSameAsStlProperty: false,
		ownerAddress: {
			addressLine1: address.streetAddress,
			addressLine2: "", // will never know
			addressLine3: "", // will never know
			town: address.addressLocality,
			postcode: address.postalCode,
			county: address.addressCountry.includes("Ireland") ? `Co. ${address.addressRegion}` : address.addressRegion
		}
	};
};
