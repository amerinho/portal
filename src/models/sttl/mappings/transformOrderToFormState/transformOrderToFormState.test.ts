import { transformOrderToFormState } from "./transformOrderToFormState";

import { formState, order } from "../mocks";

describe("mapOrder", () => {
	it("should compute expected result", () => {
		const result = transformOrderToFormState(order);
		expect(result).toMatchSnapshot();
		expect(result).toEqual(formState);
	});
});
