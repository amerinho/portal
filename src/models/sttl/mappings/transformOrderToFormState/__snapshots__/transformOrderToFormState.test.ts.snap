// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapOrder should compute expected result 1`] = `
{
  "statutory_obligations": [
    {
      "permissionStatus": "have",
    },
    {
      "permissionStatus": "have",
    },
    {
      "permissionStatus": "will_have",
    },
    {
      "permissionStatus": "have",
    },
    {
      "permissionStatus": "will_have",
    },
    {
      "permissionStatus": "will_have",
    },
  ],
  "property_address": [
    {
      "propertyAddress": {
        "addressLine1": "532 West Rocky Milton Freeway",
        "addressLine2": "",
        "addressLine3": "",
        "county": "Co. Cavan",
        "postcode": "AAA 0001",
        "town": "Aut molestiae nihil",
      },
    },
    {
      "propertyAddress": {
        "addressLine1": "634 New Freeway",
        "addressLine2": "",
        "addressLine3": "",
        "county": "Co. Carlow",
        "postcode": "AAA 0002",
        "town": "Corrupti fugiat ad",
      },
    },
    {
      "propertyAddress": {
        "addressLine1": "740 Rocky Clarendon Avenue",
        "addressLine2": "",
        "addressLine3": "",
        "county": "Co. Cork",
        "postcode": "AAA 0003",
        "town": "Ut consequat Saepe ",
      },
    },
    {
      "propertyAddress": {
        "addressLine1": "84 West Rocky Oak Street",
        "addressLine2": "",
        "addressLine3": "",
        "county": "Co. Clare",
        "postcode": "AAA 0004",
        "town": "Vitae adipisci non t",
      },
    },
    {
      "propertyAddress": {
        "addressLine1": "80 Cowley Avenue",
        "addressLine2": "",
        "addressLine3": "",
        "county": "Co. Cork",
        "postcode": "AAA 0005",
        "town": "Consequatur Ut dign",
      },
    },
    {
      "propertyAddress": {
        "addressLine1": "84 Milton Road",
        "addressLine2": "",
        "addressLine3": "",
        "county": "Co. Donegal",
        "postcode": "AAA 0006",
        "town": "Reiciendis eum moles",
      },
    },
  ],
  "property_owner_details": [
    {
      "businessName": "All Human",
      "countryOfResidence": "Ireland",
      "emailAddress": "kekyzo@mailinator.com",
      "firstName": "Cruz",
      "isAddressSameAsStlProperty": false,
      "lastName": "Obrien",
      "ownerAddress": {
        "addressLine1": "653 White Hague Extension",
        "addressLine2": "",
        "addressLine3": "",
        "county": "Co. Clare",
        "postcode": "AAA 0000",
        "town": "Aut molestiae nihil ",
      },
      "telephone": "+353851231234",
    },
    {
      "businessName": "All Human",
      "countryOfResidence": "Ireland",
      "emailAddress": "pakicihi@mailinator.com",
      "firstName": "Herman",
      "isAddressSameAsStlProperty": true,
      "lastName": "Dunlap",
      "telephone": "+353851231234",
    },
    {
      "businessName": "All Human",
      "countryOfResidence": "American Samoa",
      "emailAddress": "mejeqyhal@mailinator.com",
      "firstName": "Ian",
      "isAddressSameAsStlProperty": false,
      "lastName": "Barron",
      "ownerAddress": {
        "addressLine1": "Samoa West Avenue",
        "addressLine2": "",
        "addressLine3": "",
        "county": "Maoputasi County",
        "postcode": "ZIP 0000",
        "town": "Pago Pago",
      },
      "telephone": "+353851231234",
    },
    {
      "businessName": "All Human",
      "countryOfResidence": "Ireland",
      "emailAddress": "nove@mailinator.com",
      "firstName": "Minerva",
      "isAddressSameAsStlProperty": true,
      "lastName": "Newman",
      "telephone": "+353851231234",
    },
    {
      "businessName": "All Human",
      "countryOfResidence": "Ireland",
      "emailAddress": "vowuraqozo@mailinator.com",
      "firstName": "Blake",
      "isAddressSameAsStlProperty": true,
      "lastName": "Meadows",
      "telephone": "+353851231234",
    },
    {
      "businessName": "All Human",
      "countryOfResidence": "Ireland",
      "emailAddress": "raleso@mailinator.com",
      "firstName": "Isabella",
      "isAddressSameAsStlProperty": true,
      "lastName": "Hanson",
      "telephone": "+353851231234",
    },
  ],
  "property_type": [
    {
      "category": "room",
      "room": {
        "numberOfGuestsInPrivateRooms": 1,
        "numberOfGuestsInSharedRooms": 1,
        "numberOfPrivateRooms": 1,
        "numberOfSharedRooms": 1,
        "propertyType": "House",
      },
    },
    {
      "category": "fullProperty",
      "fullProperty": {
        "numberOfBedrooms": 1,
        "numberOfGuests": 1,
        "propertyType": "Apartment / Flat",
      },
    },
    {
      "category": "units",
      "units": [
        {
          "noOfGuests": 1,
          "propertyType": "Hostel",
        },
      ],
    },
    {
      "category": "room",
      "room": {
        "customPropertyType": "Room Special",
        "numberOfGuestsInPrivateRooms": 1,
        "numberOfGuestsInSharedRooms": 1,
        "numberOfPrivateRooms": 1,
        "numberOfSharedRooms": 1,
        "propertyType": "Other - specify",
      },
    },
    {
      "category": "fullProperty",
      "fullProperty": {
        "customPropertyType": "Entire prop Special",
        "numberOfBedrooms": 1,
        "numberOfGuests": 1,
        "propertyType": "Other - specify",
      },
    },
    {
      "category": "units",
      "units": [
        {
          "noOfGuests": 2,
          "noOfRooms": 1,
          "propertyType": "Activity or adventure centre",
        },
        {
          "noOfGuests": 2,
          "propertyType": "Hostel",
        },
        {
          "noOfGuests": 2,
          "noOfOwnDoorUnits": 1,
          "propertyType": "Apartments - student accommodation",
        },
        {
          "customPropertyType": "Multiple Other Special",
          "noOfGuests": 2,
          "number": 1,
          "propertyType": "Other - specify",
        },
      ],
    },
  ],
}
`;
